### ISSUE TITLE
version 0.8 federation-learning example seems contain an error  in validation dataset setting

### ISSUE THREAD

**Issue by wangyi160** [tf-encrypted/tf-encrypted#872]
examples/federated-learning
run.py
class ModelOwner
line 81-96

def _build_data_pipeline(self):
        """Build data pipeline for validation by model owner."""

        def normalize(image, label):
            image = tf.cast(image, tf.float32) / 255.0
            return image, label

        dataset = tf.data.TFRecordDataset(["./data/train.tfrecord"])
        dataset = dataset.map(decode)
        dataset = dataset.map(normalize)
        dataset = dataset.batch(50)
        dataset = dataset.take(1)  # keep validating on the same items
        dataset = dataset.repeat()

        iterator = dataset.make_one_shot_iterator()
        return iterator.get_next()

it seems validation should be based on test.tfrecord, but here using train.tfrecord.

Is my understanding right?

**Comment by zjn-code** [comment]
Yes, this is a mistake, and DataOwners shouldn't use same train dataset.
```
data_owners = [
        DataOwner(
            "data-owner-0", "./data/train.tfrecord", model_owner.build_update_step
        ),
        DataOwner(
            "data-owner-1", "./data/train.tfrecord", model_owner.build_update_step
        ),
        DataOwner(
            "data-owner-2", "./data/train.tfrecord", model_owner.build_update_step
        ),
    ]
```
Our new version 0.9.0 fix this problem, you could check [this](https://github.com/tf-encrypted/tf-encrypted/blob/master/examples/application/federated-learning/run.py)

**Comment by jvmncs** [comment]
fwiw this example was meant to be a demonstration of FL logic, and since the data is fake/generated I think we didn't really care to do things properly at the time.

in practice, you might want each data owner to a hold-out validation set from a partition of each of their training sets, or perhaps include an entirely new/separate validation set from the model owner. you definitely wouldn't want to create the validation set from your test set, otherwise any kind of hyperparameter tuning would bleed into your testing and lead to overly optimistic results
